lib/pymedphys/_dicom/connect/listen.py

```diff
@@ -0,0 +1,207 @@
+            (evt.EVT_ACCEPTED, self.on_association_accepted),
+            (evt.EVT_RELEASED, self.on_association_released),
+        ]
+
+        # Start listening for incoming association requests
+        self.ae.start_server((self.host, self.port), evt_handlers=handlers, block=False)
+
+    def stop(self):
+
+        if self.ae:
+            self.ae.shutdown()
+
+    def on_c_echo(self, _):
+        """Respond to a C-ECHO service request.
+        """
+        logging.debug("C-ECHO!")
+        return 0x0000
+
+    def on_association_accepted(self, _):
+        self.association_directory = None
+
+    def on_association_released(self, _):
+
+        if self.on_released_callback:
+            self.on_released_callback(self.association_directory)
+
+    def on_c_store(self, event):
+
+        dataset = event.dataset
+
+        # TODO: Should these be place in a config file?
+        mode_prefixes = {
+            "CT Image Storage": "CT",
+            "Enhanced CT Image Storage": "CTE",
+            "MR Image Storage": "MR",
+            "Enhanced MR Image Storage": "MRE",
+            "Positron Emission Tomography Image Storage": "PT",
+            "Enhanced PET Image Storage": "PTE",
+            "RT Image Storage": "RI",
+            "RT Dose Storage": "RD",
+            "RT Plan Storage": "RP",
+            "RT Structure Set Storage": "RS",
+            "Computed Radiography Image Storage": "CR",
+            "Ultrasound Image Storage": "US",
+            "Enhanced Ultrasound Image Storage": "USE",
+            "X-Ray Angiographic Image Storage": "XA",
+            "Enhanced XA Image Storage": "XAE",
+            "Nuclear Medicine Image Storage": "NM",
+            "Secondary Capture Image Storage": "SC",
+        }
+
+        try:
+            mode_prefix = mode_prefixes[dataset.SOPClassUID.name]
+        except KeyError:
+            mode_prefix = "UN"
+
+        suid = dataset.SeriesInstanceUID
+        series_dir = self.storage_directory / suid
+        series_dir.mkdir(exist_ok=True)
+
+        filename = "{0!s}.{1!s}".format(mode_prefix, dataset.SOPInstanceUID)
+        filepath = series_dir / filename
+
+        if filepath.exists:
+            # TODO Currently just logging this, since I'm unsure of the desired
+            # behaviour if the file has already been recieved.
+            logging.debug("DICOM file already exists, overwriting")
+
+        context = event.context
+        meta = Dataset()
+        meta.MediaStorageSOPClassUID = dataset.SOPClassUID
+        meta.MediaStorageSOPInstanceUID = dataset.SOPInstanceUID
+        meta.ImplementationClassUID = PYNETDICOM_IMPLEMENTATION_UID
+        meta.TransferSyntaxUID = context.transfer_syntax
+
+        # The following is not mandatory, set for convenience
+        meta.ImplementationVersionName = PYNETDICOM_IMPLEMENTATION_VERSION
+        file_ds = FileDataset(
+            filepath.as_posix(), {}, file_meta=meta, preamble=b"\0" * 128
```

Comment:
is filepath.as_posix() portable (i.e. work on MS Windows)?

Reply:
Including the preamble means that the data is susceptible to the DICOM Preamble Vulnerability.
see
https://www.medicalconnections.co.uk/kb/DICOM-Preamble-Vulnerability 
It's OK to use it, but one must be cautious about leaving the files available to a user via the file system because malware can find the files and infect them.

Reply:
I didn't know about that preamble vulnerability. Given this codes aspiration to one day be a part of a hosted API this is a key pick up. Thanks @sjswerdloff. Really appreciate your eyes on this.

Reply:
That is interesting, good to know!

I found this way of setting the preamble: `b"\0" * 128`  from the pydicom docs. This seems to write 128 zero bytes to the preamble, so would that avoid the vulnerability you mentioned? So would this overwrite any (potentially malicious) preamble in the dataset received? Or am I misunderstanding something here?

Reply:
As for the as_posix function, I do believe that it works on Windows, but there maybe be a more appropriate way to convert a platlib.Path back to a string. Do either of you have a better suggestion to use in scenarios like this?

Reply:
str(path) does the trick :). But actually, it's not needed here. We've pinned pydicom to >= 2.0.0 in pymedphys, and that version of pydicom supports pathlib objects natively :).

Reply:
Even better!

Reply:
> That is interesting, good to know!
> 
> I found this way of setting the preamble: `b"\0" * 128` from the pydicom docs. This seems to write 128 zero bytes to the preamble, so would that avoid the vulnerability you mentioned? So would this overwrite any (potentially malicious) preamble in the dataset received? Or am I misunderstanding something here?

When receiving a file over the DICOM wire protocol, there is no pre-amble, so data received that way is sterile.
When you write a file with a preamble with 0's, the file is momentarily safe.
But once it is on the file system, malware that seeks DICOM files can modify the preamble to contain a virus.
Should a user at that point double click the DICOM file (thinking that they are going to launch a default application/viewer), the executable (virus) will start running.

So the take-home is that any software reading in a DICOM file with a pre-amble should sterilise it (either by overwriting with zeros, or by removing the preamble).  And any system that stores DICOM files should consider only storing them without the preamble, and either periodically or on activation, sweep through to make sure that none of the DICOM files have preambles (or alternatively, ensure they are sterile, e.g. have all 0's).

For the DICOM Standard organisations point of view see:
https://www.dicomstandard.org/docs/librariesprovider2/dicomdocuments/wp-cotent/uploads/2019/05/faq-dicom-128-byte-preamble-posted1-1.pdf

Reply:
That's a great explanation, thank you.

Reply:
In multiple locations, it does seem like the accepted solution here is to write a preamble of 128 0x00 bytes:

* ["And when writing the DICOM instance back out, DicomObjects creates an empty 128 bytes for the preamble."](https://www.medicalconnections.co.uk/kb/DICOM-Preamble-Vulnerability#workstation-reading-applications:~:text=And%20when%20writing%20the%20DICOM%20instance,empty%20128%20bytes%20for%20the%20preamble.)
* https://pydicom.github.io/pydicom/dev/reference/generated/pydicom.filewriter.dcmwrite.html#pydicom.filewriter.dcmwrite:~:text=128%200x00%20bytes
* ["DICOM receivers commonly add a blank preamble (128 bytes of 00H) onto images they receive."](https://www.dicomstandard.org/docs/librariesprovider2/dicomdocuments/wp-cotent/uploads/2019/05/faq-dicom-128-byte-preamble-posted1-1.pdf)

Also, at the end of the day, the goal is to not be the vector by which an attack comes. So when loading DICOM files from disk, replace the preamble with 00H if one is transfering the file elsewhere as a file. If sending via DICOM network transmissions the preamble is not included anyway.

If the user already has malware executing on their machine manipulating DICOM preambles the goal is to not let other users be attacked via our software with that preamble. We can't protect the user themselves, they're already compromised.

Therefore, saving the zero bytes preamble is being DICOM compliant as far as I can tell. And it is not introducing a transmission attack vector.

Reply:
Another note on this topic. Including no header as opposed to 0's doesn't actually offer any protection. An executable on the user's system can happily take a DICOM file with no preamble and add a preamble to it. So the only difference of no preamble vs preamble I believe is one compliant to the DICOM file spec, and the other isn't. ...so a preamble of all 0's, from what I can tell is not only a good way to go, it appears to be "the" way to go.